src/circuit.rs

```diff
@@ -2,61 +2,759 @@
+                        config.advices[0],
+                        0,
+                        || scalar_val.ok_or(plonk::Error::SynthesisError),
+                    )?;
+                    Ok(CellValue::new(cell, scalar_val))
+                },
+            )?;
+
+            // Multiply scalar by NullifierK
+            // `product` = [poseidon_hash(nk, rho_old) + psi_old] NullifierK.
+            //
+            let product = {
+                let nullifier_k = FixedPointBaseField::from_inner(ecc_chip.clone(), NullifierK);
+                nullifier_k.mul(
+                    layouter.namespace(|| "[poseidon_output + psi_old] NullifierK"),
+                    scalar,
+                )?
+            };
+
+            // Add cm_old to multiplied fixed base to get nf_old
+            // cm_old + [poseidon_output + psi_old] NullifierK
+            let nf_old = cm_old
+                .add(layouter.namespace(|| "nf_old"), &product)?
+                .extract_p();
+
+            // Constrain nf_old to equal public input
+            layouter.constrain_instance(nf_old.inner().cell(), config.primary, NF_OLD)?;
+
+            nf_old
+        };
+
+        // Spend authority
+        {
+            // alpha_commitment = [alpha] SpendAuthG
+            let (alpha_commitment, _) = {
+                let spend_auth_g = OrchardFixedBasesFull::SpendAuthG;
+                let spend_auth_g = FixedPoint::from_inner(ecc_chip.clone(), spend_auth_g);
+                spend_auth_g.mul(layouter.namespace(|| "[alpha] SpendAuthG"), self.alpha)?
+            };
+
+            // [alpha] SpendAuthG + ak
+            let rk = alpha_commitment.add(layouter.namespace(|| "rk"), &ak)?;
+
+            // Constrain rk to equal public input
+            layouter.constrain_instance(rk.inner().x().cell(), config.primary, RK_X)?;
+            layouter.constrain_instance(rk.inner().y().cell(), config.primary, RK_Y)?;
+        }
+
+        // Diversified address integrity.
+        let pk_d_old = {
+            let commit_ivk_config = config.commit_ivk_config.clone();
+
+            let ivk = {
+                let rivk = self.rivk.map(|rivk| rivk.inner());
+
+                commit_ivk_config.assign_region(
+                    config.sinsemilla_chip_1(),
+                    ecc_chip.clone(),
+                    layouter.namespace(|| "CommitIvk"),
+                    *ak.extract_p().inner(),
+                    nk,
+                    rivk,
+                )?
+            };
+
+            // [ivk] g_d_old
+            // The scalar value is passed through and discarded.
+            let (derived_pk_d_old, _ivk) =
+                g_d_old.mul(layouter.namespace(|| "[ivk] g_d_old"), ivk.inner())?;
+
+            // Constrain derived pk_d_old to equal witnessed pk_d_old
+            // This addresses the case where ivk = ⊥ , since ⊥ maps to 0 and variable-base
+            // scalar multiplication maps [0]B to (0, 0).
```

Comment:
```suggestion
```
The comment is maybe a bit misleading, since ivk = ⊥ in the spec is modelling the case where Commit<sup>ivk</sup> has hit an exceptional case, and so we don't know what the circuit will actually compute for ivk (we're relying on this happening with negligible probability).